src/main/java/Core/Minimax.java

```diff
@@ -7,70 +7,77 @@
     }
 
-    public int findSquareChoice(Grid grid, Mark optimisingPlayerMark, Integer depth) {
-
-        ArrayList<Integer> emptyGridSquares = grid.emptySquareIndices();
+    public int findGridChoice() {
+        ArrayList<Integer> emptyGridSquares = firstGrid.emptySquareIndices();
         ArrayList<Integer> scores = new ArrayList<>();
 
-        emptyGridSquares
-                .stream()
-                .map(emptySquare -> emulateGrid(grid, emptySquare, optimisingPlayerMark))
-                .forEach(emulatedGrid ->
-                {
-                    Integer emulatedGridScore = findGridScore(emulatedGrid, optimisingPlayerMark, depth);
-                    scores.add(emulatedGridScore);
-                });
+        for (Integer emptySquare : emptyGridSquares) {
+            Grid emulatedGrid = emulateGrid(firstGrid, emptySquare, maximisingPlayerMark);
+            Integer score = minimax(emulatedGrid, minimisingPlayerMark, 0, Integer.MIN_VALUE, Integer.MAX_VALUE);
+            scores.add(score);
+        }
 
-        Integer relevantIndex;
-        if (grid != firstGrid) {
-            if (optimisingPlayerMark == minimaxPlayerMark) {
-                relevantIndex = findMaxIndex(scores);
-            } else {
-                relevantIndex = findMinIndex(scores);
-            }
-            return scores.get(relevantIndex);
+        Integer maxScoreIndex = findMaxIndex(scores);
+        return emptyGridSquares.get(maxScoreIndex);
+    }
+
+    private int minimax(Grid grid, Mark optimisingPlayerMark, Integer depth, Integer alpha, Integer beta) {
+
+        if (grid.isGameOver()) {
+            return scoreForTerminalGameState(grid, depth);
         } else {
-            relevantIndex = findMaxIndex(scores);
-            return emptyGridSquares.get(relevantIndex);
+            Integer bestVal;
+            if (isMaximisingPlayer(optimisingPlayerMark)) bestVal = Integer.MIN_VALUE;
+            else bestVal = Integer.MAX_VALUE;
+
+            ArrayList<Integer> emptyGridSquares = grid.emptySquareIndices();
+            for (Integer emptySquare : emptyGridSquares) {
+                Grid emulatedGrid = emulateGrid(grid, emptySquare, optimisingPlayerMark);
+                Integer value = minimax(emulatedGrid, oppositeOptimisingMark(optimisingPlayerMark), depth + 1, alpha, beta);
+                if (isMaximisingPlayer(optimisingPlayerMark)) {
+                    bestVal = Math.max(bestVal, value);
+                    alpha = Math.max(alpha, bestVal);
+                } else {
+                    bestVal = Math.min(bestVal, value);
+                    beta = Math.min(beta, bestVal);
+                }
+                if (beta <= alpha) break;
+            }
+            return bestVal;
         }
     }
 
+    private boolean isMaximisingPlayer(Mark optimisingPlayerMark) {
+        return optimisingPlayerMark == maximisingPlayerMark;
+    }
+
     private Grid emulateGrid(Grid grid, Integer emptySquare, Mark mark) {
```

Comment:
What is this doing? Taking a grid, making a move and returning the updated grid. `emulateGrid` feels a bit generic. Wouldn't something like `playMove` be more descriptive?